### PR TITLE
fix: add option to reconnect if connack has an error code

### LINKED ISSUE #1947: [Bug]: Does not start reconnect timer on connection error
### MQTTjs Version

5.10.1

### Broker

emqx

### Environment

NodeJS

### Description

If a connection attempt is met with a connack packet with an error (e.g. an auth error), no reconnect is attempted.

With expiring auth keys, such connack errors are expected. Our app is setup to trigger auth refresh on those errors using the suggested methods from other issues, but the client never actually sets up the reconnect timer in this case.

### Minimal Reproduction

```typescript
const mqtt = require('mqtt');

let auth: { username: string; password: string; };
// auth setup elided. Assume it happens to start with an expired token that needs to be renewed, but we don't know that yet.
const client = mqtt.connect('ssl://our.mqtt.broker:443', {
	...auth,
	clientId,
	log: mqttLogFunc,
	protocolVersion: 5,
	reconnectPeriod: 1000,
	keepalive: 60,
	connectTimeout: 5_000,
});
client.on('connect', () => console.log('connected'));
client.on('error', () => {
  rewnewAuth(client); // updates client.options following e.g. https://github.com/mqttjs/MQTT.js/issues/533#issuecomment-625388090
});
```

I _expect_ this to keep retrying, but it does not.

### Debug logs

These logs are from a test hitting our real broker with bogus creds. The test is nearly identical to the repro above. I have intentionally provided a wrong `password` in the test to reproduce the issue. I have also elided the username, client id, and broker hostname.

```
  mqttjs connecting to an MQTT broker... +0ms
  mqttjs:client MqttClient :: version: 5.10.1 +0ms
  mqttjs:client MqttClient :: environment node +1ms
  mqttjs:client MqttClient :: options.protocol ssl +0ms
  mqttjs:client MqttClient :: options.protocolVersion 5 +0ms
  mqttjs:client MqttClient :: options.username CENSORED +0ms
  mqttjs:client MqttClient :: options.keepalive 60 +0ms
  mqttjs:client MqttClient :: options.reconnectPeriod 1000 +0ms
  mqttjs:client MqttClient :: options.rejectUnauthorized undefined +0ms
  mqttjs:client MqttClient :: options.properties.topicAliasMaximum undefined +0ms
  mqttjs:client MqttClient :: clientId CENSORED +0ms
  mqttjs:client MqttClient :: setting up stream +1ms
  mqttjs:client connect :: calling method to clear reconnect +1ms
  mqttjs:client _clearReconnect : clearing reconnect timer +0ms
  mqttjs:client connect :: using streamBuilder provided to client to create stream +0ms
  mqttjs calling streambuilder for ssl +19ms
  mqttjs:tls port 443 host our.mqtt.broker rejectUnauthorized %b true +0ms
  mqttjs:client connect :: pipe stream to writable stream +1ms
  mqttjs:client connect: sending packet `connect` +1ms
  mqttjs:client _writePacket :: packet: {
  mqttjs:client   cmd: 'connect',
  mqttjs:client   protocolId: 'MQTT',
  mqttjs:client   protocolVersion: 5,
  mqttjs:client   clean: true,
  mqttjs:client   clientId: 'CENSORED',
  mqttjs:client   keepalive: 60,
  mqttjs:client   username: 'CENSORED',
  mqttjs:client   password: 'fake',
  mqttjs:client   properties: undefined
  mqttjs:client } +0ms
  mqttjs:client _writePacket :: emitting `packetsend` +0ms
  mqttjs:client _writePacket :: writing to stream +0ms
  mqttjs:client _writePacket :: writeToStream result true +26ms
  mqttjs:client writable stream :: parsing buffer +267ms
  mqttjs:client parser :: on packet push to packets array. +1ms
  mqttjs:client work :: getting next packet in queue +0ms
  mqttjs:client work :: packet pulled from queue +0ms
  mqttjs:client _handlePacket :: emitting packetreceive +0ms
  mqttjs:client _handleConnack +0ms
// test logs ErrorWithReasonCode: Connection refused: Not authorized at this point
```
MQTT logs nothing more after this, since ... it's not doing anything.

### PR DESCRIPTION
**Pull request by mgabeler-lee-6rs**
fixes #1947

- [x] Not sure about the naming of the new option
- [x] New option defaulting disabled to avoid surprise in consumers, but there's an argument to defaulting it on
- [x] Needs tests
- [x] Add documentation

### CODE REVIEW

**Review by robertsLando — COMMENTED** [review]
The option name looks good to me. Remember to add tests and also docs

**Comment by robertsLando** [comment]
@mgabeler-lee-6rs I still don't understand why ci fails, could you try removing the test and push? Just to see if that's the reason

**Comment by mgabeler-lee-6rs** [comment]
Trying that, but somehow I suspect that the problem is more likely related to building/uploading from a fork?

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by mgabeler-lee-6rs** [comment]
Was going to suggest contacting codecov for help, but I noticed this is using the nodejs uploader, which is deprecated to the point that the repo is archived

**Comment by robertsLando** [comment]
That's what I tried here: https://github.com/mqttjs/MQTT.js/pull/1943 

But no clue why when I add that action workflow remains in wait state for no reason

**Comment by mgabeler-lee-6rs** [comment]
checks on that PR show:
> Annotations
> 1 error
> [Error](https://github.com/mqttjs/MQTT.js/actions/runs/11812893939/workflow)
codecov/codecov-action@v4 is not allowed to be used in mqttjs/MQTT.js. Actions in this workflow must be: within a repository owned by mqttjs, created by GitHub, or matching the following: mqttjs-test, npm-publish, release-it, semantic-pr, amannn/action-semantic-pull-request@*.

**Comment by robertsLando** [comment]
Oh god you thanks sooo much for that! I think I lost 4 hours trying to debug that and I totally forgot we have some restrictions on actions that can run.... Could you tell me where you got those errors? I wan't able to see them

File: test/abstract_client.ts [hunk 1]
```diff
@@ -3335,6 +3335,41 @@ export default function abstractTest(server, config, ports) {
...
+			let connackErrors = 0
+			const rcNotAuthorized = 135
+			const server2 = serverBuilder(config.protocol, (serverClient) => {
+				serverClient.on('connect', () => {
+					const rc = connackErrors === 0 ? rcNotAuthorized : 0
+					const connack =
+						version === 5 ? { reasonCode: rc } : { returnCode: rc }
+					serverClient.connack(connack)
+				})
+			})
```
[the remark this answers is not in the document]
Comment (robertsLando) [reply test/abstract_client.ts:3348]: Try to remove skip now, also I think you can remove server.close as it's handled by teardown

**Comment by robertsLando** [comment]
Ok I fixed tests and did a minor refactor, could you please fix conflicts? I moved all tests under `test/node` folder for consistency

**Comment by mgabeler-lee-6rs** [comment]
> Could you tell me where you got those errors? I wan't able to see them

I clicked into the Checks tab on the PR, and from there (on one of the older commits that failed), into the `MQTT.js Tests` on the left bar, which had a :left_speech_bubble: icon, which takes me to a page like https://github.com/mqttjs/MQTT.js/actions/runs/11968792492 which has that error at the bottom under the `Annotations` header.

> could you please fix conflicts?

On it!

**Pull request merged**